pkg/controller/statusmanager/pod_status.go

```diff
@@ -61,17 +63,135 @@ type deploymentState struct {
 	LastChangeTime time.Time
 }
 
-// SetFromPods sets the operator Degraded/Progressing/Available status, based on
+func FindStatusCondition(conditions []corev1.NodeCondition, conditionType corev1.NodeConditionType) *corev1.NodeCondition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func SetNodeCondition(conditions *[]corev1.NodeCondition, newCondition corev1.NodeCondition) {
+	if conditions == nil {
+		conditions = &[]corev1.NodeCondition{}
+	}
+	existingCondition := FindStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		*conditions = append(*conditions, newCondition)
+		return
+	}
+
+	if existingCondition.Status != newCondition.Status {
+		existingCondition.Status = newCondition.Status
+		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
+}
+
+func (status *StatusManager) setNodeNetworkUnavailable(nodeName string, ready bool, reason string, message string) {
+	if ready == true {
+		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to true for node %s", nodeName))
+	} else {
+		log.Info(fmt.Sprintf("Setting status NetworkUnavailable to false for node %s", nodeName))
+	}
+	node := &corev1.Node{}
+	err := status.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Failed to get node %s", nodeName))
+		return
+	}
+	networkUnavailableCondition := corev1.ConditionFalse
+	if ready == false {
+		networkUnavailableCondition = corev1.ConditionTrue
+		reason += "NotReady"
+	} else {
+		reason += "Ready"
+		message = "NSX node agent is running"
+	}
+	condition := corev1.NodeCondition{
+		Type:               "NetworkUnavailable",
+		Status:             networkUnavailableCondition,
+		LastHeartbeatTime:  metav1.Now(),
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            message,
+	}
+	SetNodeCondition(&node.Status.Conditions, condition)
```

Comment:
Looks like you already check existing condition in `SetNodeCondition`, you can leverage a return value from `SetNodeCondition` to avoid updating node condition

Reply:
I did a test, whenthe node was updated via `status.client.Status().Update(context.TODO(), node)` with the unchanged condition, the lastTransitionTime in that node condition won't be changed. e.g. in operator log we can see (timestamp 1607070276 indicates UTC time 2020/12/4 08:24:36):
```
{"level":"info","ts":1607070276.3644621,"logger":"status_manager","msg":"Setting status NetworkUnavailable to flase for node wdc-rdops-vm10-dhcp-7-184"}
{"level":"info","ts":1607070276.4882815,"logger":"status_manager","msg":"Updated node condition"}
```
but after that the node condition is still:
```
lastHeartbeatTime: "2020-11-18T07:19:46Z"
lastTransitionTime: "2020-12-04T08:21:50Z"
message: NSX node agent is running
reason: NSXNodeAgentReady
status: "False"
type: NetworkUnavailable
```
So I think the `status.client.Status().Update()` can avoid the duplicate update by itself, we don't need to do more things.

Reply:
But you have an extra API call, right?

Reply:
Make sense. You remind me it's possible that the update request was sent while the api server didn't modify the timestamp.. Will investigate further and confirm if we need to skip the update from our side, the possible improvement will be in the next patch.